### ISSUE TITLE
On the way to dBFT 3.0

### ISSUE THREAD

**Issue by vncoelho** [neo-project/neo#1337]
These genius are doing great works on libraries for ensuring more proofs for dBFT, @igormcoelho and @vang1ong7ang, respectivelly: https://github.com/NeoResearch/libbft & https://github.com/NeoResearch/libsm

In addition, during these last days good designs were done together and a beautiful state machine graph is emerging!
Stay tuned. aheuahuea

![graphviz-dbft-v3-cancel-phase](https://user-images.githubusercontent.com/17395762/70364292-a90ca680-186a-11ea-84e9-a77e3853cfd6.jpg)
[graphviz-dbft-v3-cancel-phase.pdf](https://github.com/neo-project/neo/files/3934433/graphviz-dbft-v3-cancel-phase.pdf)

Currently, the discussion surrounding the version number emerged, if it should be a major version or a minor version.
Perhaps, it is a Major Version because of the following features (under update):
* New phase for choosing a redundant primary;
* Introduction of Permanent and Active backups;
* New governance for being a CN in terms of:
    * Staking instead of paying a `SystemFee`;
    * Automatically being a `CN` as an Active one.

**Comment by eryeer** [comment]
@vncoelho  I'm very interested in your dBFT 3.0 design. Do you have any doc about the specific descrption of dbft 3.0, it is difficult for me to understand just by looking at this UML diagram, haha.

**Comment by vncoelho** [comment]
Hi @eryeer,

In fact, that is true, the UML is not easy to understand. We talked about this during @vang1ong7ang journey here in Brazil.
Furthermore, he was the one that was engaged in designing this UML. In this sense, perhaps you can also talk directly to him.

For now, we are still working on the mathematical proofs of dBFT 2.0 and libraries for testing the design of dBFT 3.0.

We prefer to keep description simple for now for avoiding confusion.
Let's stay tuned, we believe that this month of January will be good for working on that.

**Comment by igormcoelho** [comment]
@eryeer @erikzhang @shargon @wanglongfei88 please find attached pre-release report on dBFT 3.0 studies. We tried to keep it short, but it already takes some pages... so feel free to start from last page (Highlights), which is just a single page, and then get complete report. It's not  completed yet, but it already aggregates several interesting community discussions from past few months/years.

[Report on dBFT 3.0 - Pre-Release 2019-01-19](https://github.com/NeoResearch/dbft3-spork-detect/blob/v0.3/docs/dBFT3_0-Report-NeoResearch-prerelease-19-01-19.pdf)

In short: we believe that multiple proposals are feasible on practice, reducing the burden of a single Primary, by allowing other "guest" nodes (possibly community nodes with high stakes locked on NEO) to also participate (eventually) and earn fees.
We start modeling a double primary mechanism, which is not yet proposed in academic literature, to the best of our knowledge. We have many citations not yet included on text (but present on our minds), so it's likely we forgot to mention many things, sorry for that. But it's a solid start. Enjoy ;)

**Comment by vncoelho** [comment]
We were today able to find strong mathematically evidences that double proposals will not produce sporks:
Here is the current mathematical model: https://github.com/NeoResearch/milp_bft_failures_attacks/blob/master/dbft3.0/MILP_BFTConsensus_dBFT3.0_2P.mod

We considered the model with 2 Primaries, which we believe will benefit the ecosystem in different aspects.
